schema/oneof_string_test.go

```diff
@@ -222,3 +243,355 @@ func TestOneOfStringCompatibilityMapValidation(t *testing.T) {
+		"d_type":        "A",
+		"other_field_a": "test",
+	}
+	// Since this is not struct-mapped, unserializedData is a map.
+	unserializedData := assert.NoErrorR[any](t)(oneofSchema.Unserialize(serializedData))
+	reserializedData := assert.NoErrorR[any](t)(oneofSchema.Serialize(unserializedData))
+	assert.Equals[any](t, reserializedData, serializedData)
+
+}
+
+type inlinedTestIntDiscriminatorA struct {
+	DType       int    `json:"d_type"`
+	OtherFieldA string `json:"other_field_a"`
+}
+
+var inlinedTestIntDiscriminatorAProperties = map[string]*schema.PropertySchema{
+	"d_type": schema.NewPropertySchema(
+		schema.NewIntSchema(nil, nil, nil),
+		nil, true, nil, nil, nil,
+		nil, nil,
+	),
+	"other_field_a": schema.NewPropertySchema(
+		schema.NewStringSchema(nil, nil, nil),
+		nil, true, nil, nil, nil,
+		nil, nil,
+	),
+}
+
+var inlinedTestNoDiscriminatorBProperties = map[string]*schema.PropertySchema{
+	"other_field_b": schema.NewPropertySchema(
+		schema.NewStringSchema(nil, nil, nil),
+		nil, true, nil, nil, nil,
+		nil, nil,
+	),
+}
+
+var inlinedTestIntDiscriminatorAMappedSchema = schema.NewStructMappedObjectSchema[inlinedTestIntDiscriminatorA](
+	"inlined_int_A",
+	inlinedTestIntDiscriminatorAProperties,
+)
+
+var inlinedTestIntDiscriminatorASchema = schema.NewObjectSchema(
+	"inlined_int_A",
+	inlinedTestIntDiscriminatorAProperties,
+)
+
+var inlinedTestIntDiscriminatorBSchema = schema.NewObjectSchema(
+	"inlined_int_B",
+	inlinedTestNoDiscriminatorBProperties,
+)
+
+func TestOneOf_Error_SubtypeHasInvalidDiscriminatorType(t *testing.T) {
+	testSchema := schema.NewOneOfStringSchema[any](map[string]schema.Object{
+		"A": inlinedTestIntDiscriminatorAMappedSchema,
+		"B": inlinedTestObjectBMappedSchema,
+	}, "d_type", true)
+
+	assert.Panics(t, func() {
+		schema.NewScopeSchema(schema.NewObjectSchema("test",
+			map[string]*schema.PropertySchema{
+				"test": schema.NewPropertySchema(
+					testSchema,
+					nil, true, nil, nil,
+					nil, nil, nil),
+			}))
+	})
+}
+
+func TestOneOf_Error_InlineSubtypeMissingDiscriminator(t *testing.T) {
+	testSchema := schema.NewOneOfIntSchema[any](map[int64]schema.Object{
+		1: inlinedTestIntDiscriminatorASchema,
+		2: inlinedTestIntDiscriminatorBSchema,
+	}, "d_type", true)
+
+	assert.Panics(t, func() {
+		schema.NewScopeSchema(schema.NewObjectSchema("test",
+			map[string]*schema.PropertySchema{
+				"test": schema.NewPropertySchema(
+					testSchema,
+					nil, true, nil, nil,
+					nil, nil, nil),
+			}))
+	})
+}
+
+func TestOneOf_Error_SubtypeHasDiscriminator(t *testing.T) {
+	testSchema := schema.NewOneOfStringSchema[any](map[string]schema.Object{
+		"A": inlinedTestIntDiscriminatorASchema,
+		"B": nonInlinedTestObjectBSchema,
+	}, "d_type", false)
+
+	assert.Panics(t, func() {
+		schema.NewScopeSchema(schema.NewObjectSchema("test",
+			map[string]*schema.PropertySchema{
+				"test": schema.NewPropertySchema(
+					testSchema,
+					nil, true, nil, nil,
+					nil, nil, nil),
+			}))
+	})
+}
```

Comment:
Did you plan on using one of the new panics assertions? All you need to do is upgrade to the latest assertion library version. https://github.com/arcalot/go-assert/releases/tag/v1.8.0

Reply:
I did. I'll get on it.

Reply:
push the change